### DISCUSSION TITLE
How to control sorting of topics?

### DISCUSSION THREAD

**Discussion by emin63** [magit/forge#393] [Support] [answered]
How can I control sorting of topics/issues?

It seems like there is a `forge-topic-list-columns` variable which includes a reference to `forge-topic-list-sort-by-number` but I can't find documentation about how to use that. 

Thanks.

**Comment by tarsius** [comment]
The value of `tabulated-list-sort-key` is set in `forge-topic-list-mode` (which see). You should be able to use `forge-topic-list-mode-hook` to pick another key.

**Reply by emin63** [reply]
Thank you, that is helpful.

What I would really like to do is sort on the label (or anything else which I can use to prioritize topics/issues). Do you know how I can get the labels into `forge-topic-list-columns` so I can try to use it in `tabulates-list-sort-key`?

Thanks again.

**Reply by tarsius** [reply]
I've just added the commands `forge-list-labeled-pullreqs` and `forge-list-labeled-issues`. There are no default key bindings for now and you will have to type the label name (with completion).

**Reply by emin63** [reply]
Thank you for the quick response. I'm not sure how to use either of those to achieve what I want, however.

What I want is to have the `forge-topic-issues` command (bound to `' l i` in my system) to show the labels so I can sort on labels.

After poking around, it seemed like maybe I could add a new column by changing the `forge-topic-list-columns` variable to add a "Labels" column but I'm not sure of the right way to do that.

For example, I would like to do something like the following:
```
(setq forge-topic-list-columns
      '(("#" 5 forge-topic-list-sort-by-number (:right-align t) number nil)
	("Label" 35 t nil labels nil)  ;; this line fails
	("Title" 35 t nil title nil)))
```
but I don't know what to put in for `labels` (with the above I get the error `EmacSQL had an unhandled condition: "no such column: label"`).

**Reply by tarsius** [reply]
You are not saying why you want to sort by labels but I assume it is so that topics (i.e. rows) that have the same can be made to appear next to each other. Because a topic can have more than one label no amount of sorting would cause e.g. "`albatross`, `bug`" and "`zebra`, `bug`" to appear close to each other, without any non-`bug`s in between.

Filtering however works, which is why I have added the new `forge-list-labeled-pullreqs` and `forge-list-labeled-issues` commands.

> but I don't know what to put in for `labels` (with the above I get the error `EmacSQL had an unhandled condition: "no such column: label"`).

You can use the name of the database-column / object-slot if a database-column is actually being used to represent the object-slot. That is only the case the values are atomic. E.g. the name of a repository is an atomic value, the labels however are not.

In other places where I used `tabulated-list-mode` I also made it possible to use arbitrary functions instead of slot symbols. Apparently that doesn't work here yet. Otherwise the same thing could be done here as when the topics are inserted into the status buffer (see `forge--format-topic-labels` as called by `forge--insert-topic-contents` as called by ...

I agree it would be nice to show the labels in the `tabulated-list` view also. I have added this to my todo list but as a low-priority.

**Comment by emin63** [comment, marked as answer]
TL;DR: Your hints helped me figure it out. See https://github.com/emin63/magit-priority-example for the code I wrote (based on your hints) to achieve the goal and explain why. If you think this is generally useful, I would be happy to create a pull request to merge it into the main code base. Longer discussion below.

> 
> You are not saying why you want to sort by labels but I assume it is so that topics (i.e. rows) that have the same can be made to appear next to each other. Because a topic can have more than one label no amount of sorting would cause e.g. "`albatross`, `bug`" and "`zebra`, `bug`" to appear close to each other, without any non-`bug`s in between.

Good point. What I wanted was filtering based on a pattern (e.g., a SQL query with the `LIKE` syntax). That way I can create labels like `priority:1`, `priority:2`, and so on and then get all issues with labels matching `priority:%` sorted in ascending order to see prioritized issues.

> 
> Filtering however works, which is why I have added the new `forge-list-labeled-pullreqs` and `forge-list-labeled-issues` commands.

Thanks. Your `forge-list-labeled-issues` function didn't quite do what I wanted but it gave me the hints to create the desired results. See https://github.com/emin63/magit-priority-example for a proposed improved version of `forge-list-labeled-issues`.

> In other places where I used `tabulated-list-mode` I also made it possible to use arbitrary functions instead of slot symbols. Apparently that doesn't work here yet. Otherwise the same thing could be done here as when the topics are inserted into the status buffer (see `forge--format-topic-labels` as called by `forge--insert-topic-contents` as called by ...

> I agree it would be nice to show the labels in the `tabulated-list` view also. I have added this to my todo list but as a low-priority.

OK. I am not quite familiar enough with elisp to figure out how to make it so that labels can be used with `tabulated-list-mode` myself.  Instead, I did a bit of a hack to put the desired label in front of the issue title.

I am ~closing this issue~ marking as answered since I think the goal is accomplished. Please let me know either in this issue or on https://github.com/emin63/magit-priority-example or however you find convenient if I should create a pull request or if you think something else needs to be done to merge this code into your repo.

Thanks again for your help.

**Comment by tarsius** [comment]
Create that you came up with a solution that works for you. Also I now understand better what functionality you want.

Instead of merging an improved version of your implementation I still intend to port the abstraction I use in other `tabulated-list`-using packages and then implement the mentioned features on top of that. Your implementation gave me some ideas and important hints, so it is useful even though the code won't make it in.

**Reply by emin63** [reply]
Sounds good.

I understand that you might not want an elisp noob mucking up your code. I really appreciate magit and how much it improves the git/github experience. So if there is something I can do to to help or tweak that example I created, please let me know.

Best wishes,
-Emin
